### DISCUSSION TITLE
AttributeError: can't set attribute

### DISCUSSION THREAD

**Discussion by Messi0617** [pyswmm/pyswmm#463] [Q&A]
Hello, I am currently using the pyswmm package and NSGA2 algorithm to optimize the subcatchment connection in a watershed. However, I found that the subcatchment.connection can only read the outlet attribute of the subcatchments in the inp file, and it cannot set the range of values for the subcatchment connection variable. I would like to know which files in pyswmm need to be modified in order to set the range of values for the subcatchment connection, similar to setting the range of values for LID area in the inp file using LidGroups(sim)['4'][0].unit_area = X[0].(vars = {'x1': Real(bounds=(0,437.8344)), 'x2': Real(bounds=(0, 408.4665)), 'x3': Real(bounds=(0, 410.6523)), 'x4': Choice(options=[207,39,206,214,121])}) I am not very familiar with the code, so I would greatly appreciate your advice and assistance. Thank you very much.

<ipython-input-3-99b77d7bb878> in _evaluate(self, X, out, *args, **kwargs)
     27             LidGroups(sim)['5'][0].unit_area = X[1]  
     28             LidGroups(sim)['6'][0].unit_area = X[2] 
---> 29             Subcatchments(sim)['4'].connection = X[3]
     30 
     31             system_routing = SystemStats(sim)

AttributeError: can't set attribute

**Comment by bemcdonnell** [comment]
@Messi0617 sorry for the delay. It’s true - not every parameter has a “setter.”  If you would like the add it, you have to first begin in SWMM’s C code and add a new function to alter where a subcatchment directs the flow.  It wouldn’t be a huge lift to make this happen but you might be better off just editing the INP file before you open the model with pyswmm.  Just so I understand your use case better, are tou looking to change where a subcatchment directs flow during the simulation? Or do you just want to set it once before the simulation starts?

**Reply by Messi0617** [reply]
Actually, I don't just change the value of subcatchment.connection once before starting. I need to optimize the subcatchment.connection parameter by making multiple calls to SWMM. It's like optimizing the LID area size in each subcatchment through a genetic algorithm. I am optimizing the subcatchment.connection parameter for each subcatchment through a genetic algorithm.

**Reply by bemcdonnell** [reply]
@Messi0617 ok so you need to change the connection during the simulation - are you thinking to do this many times during the run? Or once?  I think what you are asking for is reasonable to implement. Can you familiarize yourself with this code base and script first? This is where we’d need to make changes to SWMM 

https://github.com/pyswmm/Stormwater-Management-Model/blob/main/src/solver/toolkit.c

**Reply by Messi0617** [reply]
What I need is to call SWMM multiple times, and change the subcatchment. connection based on the passed parameters each time SWMM is called.Currently, my programming skills are limited to writing simple functions and using existing packages. However, I am very willing to learn how to modify existing packages and use them to achieve the functionality I need. When it comes to implementing the functionality of modifying the subcatchment.connection parameter, I would greatly appreciate your guidance and assistance. If I am able to implement the functionality of modifying the subcatchment.connection parameter, I am more than willing to share this code and add it to a new version of pyswmm. Moving forward, I will familiarize myself with the codebase you shared and make any necessary modifications to it. I truly look forward to keeping in touch with you and maintaining communication.

**Reply by bemcdonnell** [reply]
@Messi0617 ok I am still having trouble understanding. Could you possibly create a flow diagram of the process that you have in mind using draw.io? I’m trying to visualize it better.  I’m hung up on “call SWMM multiple times.”  Each time pyswmm asks to modify a parameter, a call is made into SWMM.  Is this every time step?  For all subcatchments?  Sometimes people say “call SWMM” when they are simply running a simulation.  I think a flow diagram would help me understand.  What you are looking to do might be achievable in a different way.

**Reply by Messi0617** [reply]
@bemcdonnell Here is my code, which primarily uses a genetic algorithm to optimize the area and subcatchment.connection for each subcatchment. Perhaps, due to my novice programming skills, my previous explanations were not very clear, and I apologize for that. I hope this code will provide you with a clear understanding of what I am trying to accomplish. I am eagerly looking forward to receiving your suggestions and assistance.

import numpy as np

from pyswmm import Simulation, Nodes, Subcatchments, LidGroups, SystemStats
from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.core.problem import ElementwiseProblem
from pymoo.optimize import minimize
from pymoo.visualization.scatter import Scatter
from pymoo.core.variable import Choice, Real


class MyProblem(ElementwiseProblem):
    def __init__(self):
        vars = {
            'x1': Real(bounds=(0, 437.8344)),
            'x2': Real(bounds=(0, 408.4665)),
            'x3': Real(bounds=(0, 410.6523)),
            'x4': Choice(options=[207, 39, 206, 214, 121])
        }
        super().__init__(n_var=4, n_obj=2, n_constr=0,**kwargs)

    def _evaluate(self, X, out, *args, **kwargs):
        with Simulation(r'E:/benkebiyelunwen/zhengshi/newguanlv.inp') as sim:
            LidGroups(sim)['4'][0].unit_area = X["x1"] 
            LidGroups(sim)['5'][0].unit_area = X["x2"] 
            LidGroups(sim)['6'][0].unit_area = X["x3"]
            Subcatchments(sim)['4'].connection = X["x4"]

            system_routing = SystemStats(sim)
  
            for step in enumerate(sim):
                pass  

            Runoff = system_routing.runoff_stats['runoff']
            f1 = 221.64 * a_lswd
            f2 = Runoff
            out["F"] = [f1, f2]
            sim.close()

problem = MyProblem()

algorithm = NSGA2(pop_size=50)  

res = minimize(problem,
               algorithm,
               ("n_gen", 100),  
               verbose=True, 
               # save_history=True,
               seed=1)

**Reply by Messi0617** [reply]
EXPORT_TOOLKIT int swmm_setSubcatchOutConnection(int index, SM_ObjectType type, int out_index)
/// Purpose: Sets Subcatchment Connection ID Indeces for either Node or Subcatchment
    int error_code = 0;
    
    // Check if Open
    if(swmm_IsOpenFlag() == FALSE)
    {
        error_code = ERR_TKAPI_INPUTNOTOPEN;
    }
    
     // Check if Simulation is Running
    else if(swmm_IsStartedFlag() == TRUE)
    {
        error_code = ERR_TKAPI_SIM_NRUNNING;
    }
    
     // Check if object index is within bounds
    else if (index < 0 || index >= Nobjects[SUBCATCH])
    {
        error_code = ERR_TKAPI_OBJECT_INDEX;
    }
    
    else
    {
        //Set output connections for sub catchment areas based on target object type and index
        switch (type)
        {
            case NODE:
                Subcatch[index].outNode = out_index;
                Subcatch[index].outSubcatch = -1;
                break;
                
            case SUBCATCH:
                Subcatch[index].outNode = -1;
                Subcatch[index].outSubcatch = out_index;
                break;
                
            default:
                error_code = ERR_TKAPI_OUTBOUNDS;
                break;
        }
    }
    
    return error_code;
}
@bemcdonnell Based on the link you provided last week, I have added a new function. I am eagerly looking forward to your guidance and feedback.

**Reply by bemcdonnell** [reply]
@Messi0617 This code looks more-or-less correct to me.  So in order to add this feature we are going to need to go through several steps including the addition of unit tests and api documentation.  I can help you with this process but I want to bring attention to the fact that this could take several weeks.

Alternately, if you need this très vite, you could always manually change the loading node in the INP BEFORE you open it with pyswmm.  It will work the same way.  It looks like your workflow does not require that this value be changed during the running simulation (any code before the `for step in sim` section.

**Reply by Messi0617** [reply]
In fact, the code snippet provided was a simplified version. I need to run the .inp files hundreds of times. It is true that the subcatchments.connection variable does not need to be changed during the running of the .inp files, but before each run, the subcatchments.connection variable needs to be reassigned based on its range. Moreover, before each run, there are dozens of subcatchments that require modification of this variable. It is practically impossible for me to manually modify them. I indeed require this functionality and would greatly appreciate your assistance and guidance. Even if it ends up in failure, I would still benefit from the process by learning some knowledge about customization and development.
